browser/conversion-tests/conversion_test.ts

```diff
@@ -0,0 +1,1003 @@
+    WrapperValidationResults,
+    WrapperWethDepositEvent,
+    WrapperWethWithdrawalEvent,
+} from '../ts/types';
+import '../ts/wasm_exec';
+
+interface ConversionTestCase {
+    contractEventsAsync: () => Promise<WrapperContractEvent[]>;
+    getOrdersResponseAsync: () => Promise<WrapperGetOrdersResponse[]>;
+    orderEventsAsync: () => Promise<WrapperOrderEvent[]>;
+    signedOrdersAsync: () => Promise<WrapperSignedOrder[]>;
+    statsAsync: () => Promise<WrapperStats[]>;
+    validationResultsAsync: () => Promise<WrapperValidationResults[]>;
+}
+
+// The Go code sets certain global values and this is our only way of
+// interacting with it. Define those values and their types here.
+declare global {
+    // Defined in wasm_exec.ts
+    class Go {
+        public importObject: any;
+        public run(instance: WebAssembly.Instance): void;
+    }
+
+    // Define variables that are defined in `browser/go/conversion-test.go`
+    const conversionTestCases: ConversionTestCase;
+}
+
+// The interval (in milliseconds) to check whether Wasm is done loading.
+const wasmLoadCheckIntervalMs = 100;
+
+// We use a global variable to track whether the Wasm code has finished loading.
+let isWasmLoaded = false;
+const loadEventName = '0xmeshtest';
+window.addEventListener(loadEventName, () => {
+    isWasmLoaded = true;
+});
+
+// Start compiling the WebAssembly as soon as the script is loaded. This lets
+// us initialize as quickly as possible.
+const go = new Go();
+WebAssembly.instantiate(wasmBuffer, go.importObject)
+    .then(module => {
+        go.run(module.instance);
+    })
+    .catch(err => {
+        // tslint:disable-next-line no-console
+        console.error('Could not load Wasm');
+        // tslint:disable-next-line no-console
+        console.error(err);
+        // If the Wasm bytecode didn't compile, Mesh won't work. We have no
+        // choice but to throw an error.
+        setImmediate(() => {
+            throw err;
+        });
+    });
+
+/*********************** Tests ***********************/
+// tslint:disable:custom-no-magic-numbers
+// tslint:disable:no-console
+
+(async () => {
+    await waitForLoadAsync();
+    const contractEvents = await conversionTestCases.contractEventsAsync();
+    testContractEvents(contractEvents);
+    const getOrdersResponse = await conversionTestCases.getOrdersResponseAsync();
+    testGetOrdersResponse(getOrdersResponse);
+    const orderEvents = await conversionTestCases.orderEventsAsync();
+    testOrderEvents(orderEvents);
+    const signedOrders = await conversionTestCases.signedOrdersAsync();
+    testSignedOrders(signedOrders);
+    const stats = await conversionTestCases.statsAsync();
+    testStats(stats);
+    const validationResults = await conversionTestCases.validationResultsAsync();
+    testValidationResults(validationResults);
+
+    // This special #jsFinished div is used to signal the headless Chrome driver
+    // that the JavaScript code is done running.
```

Comment:
This is the same method that is used in the browser integration tests. The `ts/conversion_test.ts` file will place this `#jsFinished` `div` into the `index.html` page after all tests have finished executing. I'll add a comment, but hopefully that explains the origin of this method.

Reply:
I came up with this hack for the integration tests. It's written this way because `chromedp` has a limited API for "waiting for X to happen". It has a function called [`WaitVisible`](https://godoc.org/github.com/chromedp/chromedp#WaitVisible) which blocks until an element is visible in the DOM. It would be cleaner if we could wait for a JavaScript event or maybe a log message, but AFAICT there is no way to do that with `chromedp`.